File: typescript/src/problems101_200/construct-binary-tree-from-preorder-and-inorder-traversal.ts
import TreeNode from '../concepts/tree-node';

class Problem {
  buildTree(preorder: number[], inorder: number[]): TreeNode | undefined {
    throw new Error(`Not implemented (${preorder}, ${inorder}).`);
  }
}

class Default extends Problem {
  buildTree(preorder: number[], inorder: number[]): TreeNode | undefined {
    if (preorder.length === 0 || inorder.length === 0) {
      return undefined;
    }
    const root: TreeNode = new TreeNode(preorder[0]);
    let mid: number = -1;
    for (let i = 0; i < inorder.length; i++) {
      if (inorder[i] !== preorder[0]) {
        continue;
      }
      mid = i;
      break;
    }
    root.left =
        this.buildTree(
            preorder.slice(1, mid + 1),
            inorder.slice(0, mid)
        );
    root.right =
        this.buildTree(
            preorder.slice(mid + 1, preorder.length),
            inorder.slice(mid + 1, inorder.length)
        );
    return root;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems101_200/construct_binary_tree_from_preorder_and_inorder_traversal.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.tree_node import TreeNode


class Problem(ABC):
    @abstractmethod
    def build_tree(self, preorder: list[int], inorder: list[int]) -> TreeNode | None:
        pass


class Default(Problem):
    @override
    def build_tree(self, preorder: list[int], inorder: list[int]) -> TreeNode | None:
        if len(preorder) == 0 or len(inorder) == 0:
            return None

        root: TreeNode = TreeNode(preorder[0])
        mid: int = -1
        for i, order in enumerate(inorder):
            if order != preorder[0]:
                continue
            mid = i
            break
        root.left = \
            self.build_tree(
                preorder[1:mid + 1],
                inorder[:mid],
            )
        root.right = \
            self.build_tree(
                preorder[mid + 1:],
                inorder[mid + 1:],
            )
        return root


SOLUTIONS: list[Problem] = [Default()]
